### DISCUSSION TITLE
Launching an external terminal

### DISCUSSION THREAD

**Discussion by mark2185** [mfussenegger/nvim-dap#1006] [Q&A] [answered]
I'm trying to use a configuration from `.vscode/launch.json` to debug a TUI application written in go ([lazygit](https://github.com/jesseduffield/lazygit/)).

The issue is that it complains with "open /dev/tty: no such device or address".

This is the minimal `launch.json` I'm trying to use (it's slightly modified from the one in the repo, the `mode` is set to `"debug"`):
```json
    {
      "name": "Debug Lazygit",
      "type": "go",
      "request": "launch",
      "mode": "debug",
      "program": "main.go",
      "args": ["--debug", "--use-config-file=${workspaceFolder}/.vscode/debugger_config.yml"],
      "console": "integratedTerminal",
      "presentation": {
        "hidden": true
      }
    },
```

This is the output of the `dap-repl`:
```
[debug-adapter stderr] 2023/08/05 11:06:11 An error occurred! Please create an issue at: https://github.com/jesseduffield/lazygit/issues

*fs.PathError open /dev/tty: no such device or address
/home/mark/workspace/gits/lazygit/pkg/app/app.go:55 (0xf2cc25)
	Run: newErr := errors.Wrap(err, 0)
/home/mark/workspace/gits/lazygit/pkg/app/entry_point.go:154 (0xf30e72)
	Start: Run(appConfig, common, appTypes.NewStartArgs(cliArgs.FilterPath, parsedGitArg, integrationTest))
/home/mark/workspace/gits/lazygit/main.go:23 (0xf34312)
	main: app.Start(ldFlagsBuildInfo, nil)
/usr/lib/go/src/runtime/proc.go:250 (0x440f33)
	main: fn()
/usr/lib/go/src/runtime/asm_amd64.s:1598 (0x472e61)
	goexit: BYTE	$0x90	// NOP
```

I tried using `VSCode` to see if it works there, and the `"/dev/tty: no such device or address"` is symptomatic to a few other TUI tools using `awesome-gocui` (e.g. `lazydocker`, `dive`), but only when using `internalConsole`, with the error message being something like `"failed to initialize tcell screen: open /dev/tty: no such device or address".

If I switch to `integratedTerminal` (still in `VScode`) then it works, and if I switch to `externalTerminal` then `delve` times out trying to establish a reverse connection (I'm guessing something else is afoot, not related to this).

So I thought I might launch an external terminal in `nvim-dap` by setting `console` to `"externalTerminal"` and setting this:
```lua
require( 'dap' ).defaults.fallback.external_terminal = {
    command = '/usr/bin/kitty';
    args = {'--hold'};
}
```

Yet, the error remains.

Just to confirm that I've set the `defaults.fallback.external_terminal` correctly, I tried launching something not using `delve` so I tried a rust application [gitui](https://github.com/extrawurst/gitui/) with the following `launch.json`:
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch",
            "type": "lldb",
            "request": "launch",
            "program": "target/debug/gitui",
            "console": "externalTerminal",
        }
    ]
}
```

And it launches it in a separate `kitty` instance, as expected.

What am I missing? It's as if `externalTerminal` is being ignored when using `delve`.

Additional info:
[This](https://gist.github.com/mark2185/4f6cffebdada4239a8db570de4eb0a26) is my entire luafile for the debugger setup (I'm using lazy.nvim) minus some of the mappings.

**Comment by mfussenegger** [comment, marked as answer]
delve afaik still doesn't support either integratedTerminal or externalTerminal. That it works in vscode is because the vscode-go extension implements it on the client side.

See https://github.com/mfussenegger/nvim-dap/issues/987#issuecomment-1653126098

**Reply by mark2185** [reply]
Oh, I see. Thanks for clearing it up!

**Comment by Nicothekiller** [comment]
I know it has been some time since this was posted but i found a way to make this work. (for some context, im also using nvim-dap-ui)

simply change the debugger command for delve in the adapter from "dlv", to your console, in my case kitty. Then move dlv to the args and it should work. This also fixes the problem where the input gets stuck on the debug console, since the debug console is still on neovim. 

It should look kinda like this (this is the config i use).
```
dap.adapters.delve = {
                type = "server",
                port = "${port}",
                executable = {
                    command = "/usr/bin/kitty",
                    args = { "dlv","dap", "-l", "127.0.0.1:${port}" },
                },
            }
```

**Reply by Av1d1ty** [reply]
You saved me! Thanks a bunch.

**Reply by catgoose** [reply]
Another way to do this with the neovim terminal is:

```lua
    local function get_unused_port()
      local uv = vim.loop
      local server = uv.new_tcp()
      assert(server:bind("127.0.0.1", 0)) -- OS allocates an unused port
      local tcp_t = server:getsockname()
      server:close()
      assert(tcp_t and tcp_t.port > 0, "Failed to get an unused port")
      return tcp_t.port
    end

  dap.adapters.go = function(callback, config)
    local port = config.port or get_unused_port()
    local term_buf = vim.api.nvim_create_buf(false, true)
    vim.api.nvim_command("split")
    vim.api.nvim_command("buffer " .. term_buf)
    vim.fn.jobstart({ "dlv", "dap", "-l", "127.0.0.1:" .. port }, { term = true })
    vim.defer_fn(function()
      callback({ type = "server", host = "127.0.0.1", port = port })
    end, 100)
  end
```

**Reply by fredwbe** [reply]
THANK YOU SO MUCH i lost my mid trying to fix it
